sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/specialized/BlockBlobAsyncApiTests.java

```diff
@@ -0,0 +1,2658 @@
+        Map<String, String> tags = new HashMap<>();
+        if (key1 != null) {
+            tags.put(key1, value1);
+        }
+        if (key2 != null) {
+            tags.put(key2, value2);
+        }
+
+        blockBlobAsyncClient.uploadWithResponse(new BlockBlobSimpleUploadOptions(DATA.getDefaultInputStream(),
+            DATA.getDefaultDataSize()).setTags(tags)).block();
+
+        StepVerifier.create(blockBlobAsyncClient.getTagsWithResponse(new BlobGetTagsOptions()))
+            .assertNext(r -> {
+                assertResponseStatusCode(r, 200);
+                assertEquals(tags, r.getValue());
+            })
+            .verifyComplete();
+    }
+
+    @DisabledIf("com.azure.storage.blob.BlobTestBase#olderThan20191212ServiceVersion")
+    @ParameterizedTest
+    @MethodSource("com.azure.storage.blob.BlobTestBase#allConditionsSupplier")
+    public void uploadAC(OffsetDateTime modified, OffsetDateTime unmodified, String match, String noneMatch,
+                         String leaseID, String tags) {
+        Map<String, String> t = new HashMap<>();
+        t.put("foo", "bar");
+        blockBlobAsyncClient.setTags(t).block();
+        match = setupBlobMatchCondition(blockBlobAsyncClient, match);
+        leaseID = setupBlobLeaseCondition(blockBlobAsyncClient, leaseID);
+        BlobRequestConditions bac = new BlobRequestConditions()
+            .setLeaseId(leaseID)
+            .setIfMatch(match)
+            .setIfNoneMatch(noneMatch)
+            .setIfModifiedSince(modified)
+            .setIfUnmodifiedSince(unmodified)
+            .setTagsConditions(tags);
+
+        assertAsyncResponseStatusCode(blockBlobAsyncClient.uploadWithResponse(DATA.getDefaultFlux(),
+            DATA.getDefaultDataSize(), null, null, null, null, bac), 201);
+    }
+
+    @DisabledIf("com.azure.storage.blob.BlobTestBase#olderThan20191212ServiceVersion")
+    @ParameterizedTest
+    @MethodSource("com.azure.storage.blob.BlobTestBase#allConditionsFailSupplier")
+    public void uploadACFail(OffsetDateTime modified, OffsetDateTime unmodified, String match, String noneMatch,
+                             String leaseID, String tags) {
+        noneMatch = setupBlobMatchCondition(blockBlobAsyncClient, noneMatch);
+        setupBlobLeaseCondition(blockBlobAsyncClient, leaseID);
+        BlobRequestConditions bac = new BlobRequestConditions()
+            .setLeaseId(leaseID)
+            .setIfMatch(match)
+            .setIfNoneMatch(noneMatch)
+            .setIfModifiedSince(modified)
+            .setIfUnmodifiedSince(unmodified)
+            .setTagsConditions(tags);
+
+        StepVerifier.create(blockBlobAsyncClient.uploadWithResponse(DATA.getDefaultFlux(), DATA.getDefaultDataSize(),
+            null, null, null, null, bac))
+            .verifyErrorSatisfies(r -> {
+                BlobStorageException e = assertInstanceOf(BlobStorageException.class, r);
+                assertTrue(e.getErrorCode() == BlobErrorCode.CONDITION_NOT_MET
```

Comment:
Something that we could make better in the future for debuggability, assertTrue (and assertFalse) will only tell you that the boolean didn't match the expected, these methods also have a parameter for a String or Supplier<String> where we could have it say what the error code was when it doesn't match.

This doesn't need to be done in this PR and is a wider issue in the repo.